src/Utilities/cachingoptimizer.jl

```diff
@@ -428,52 +428,78 @@ function MOI.supports(m::CachingOptimizer, attr::MOI.AbstractModelAttribute)
+                                MOI.AbstractVariableAttribute,
+                                MOI.AbstractConstraintAttribute})
 
-function MOI.get(m::CachingOptimizer, attr::Union{MOI.AbstractVariableAttribute,MOI.AbstractConstraintAttribute}, index::MOI.Index)
-    if MOI.canget(m.model_cache, attr, typeof(index))
-        return MOI.get(m.model_cache, attr, index)
-    elseif m.state == AttachedOptimizer && MOI.canget(m.optimizer, attr, typeof(index))
-        return attribute_value_map(m.optimizer_to_model_map,MOI.get(m.optimizer, attr, m.model_to_optimizer_map[index]))
-    end
-    error("Attribute $attr not accessible")
-end
+Return a `Bool` indicating whether the value of the attribute is determined
+during [`MOI.optimize!`](@ref) hence is part of the result of the optimization.
+"""
+function is_result_attribute end
 
-function MOI.get(m::CachingOptimizer, attr::Union{MOI.AbstractVariableAttribute,MOI.AbstractConstraintAttribute}, indices::Vector{<:MOI.Index})
-    if MOI.canget(m.model_cache, attr, eltype(indices))
-        return MOI.get(m.model_cache, attr, indices)
-    elseif m.state == AttachedOptimizer && MOI.canget(m.optimizer, attr, eltype(indices))
-        return attribute_value_map.(Ref(m.optimizer_to_model_map),MOI.get(m.optimizer, attr, getindex.(Ref(m.model_to_optimizer_map),indices)))
-    end
-    error("Attribute $attr not accessible")
+function is_result_attribute(::Union{MOI.AbstractModelAttribute,
+                                     MOI.AbstractVariableAttribute,
+                                     MOI.AbstractConstraintAttribute})
+    return false
 end
-
-function MOI.canget(m::CachingOptimizer, attr::MOI.AbstractModelAttribute)
-    MOI.canget(m.model_cache, attr) && return true
-    if m.state == AttachedOptimizer
-        MOI.canget(m.optimizer, attr) && return true
+function is_result_attribute(::Union{MOI.ObjectiveValue,
+                                     MOI.ObjectiveBound,
+                                     MOI.RelativeGap,
+                                     MOI.SolveTime,
+                                     MOI.SimplexIterations,
+                                     MOI.BarrierIterations,
+                                     MOI.NodeCount,
+                                     MOI.RawSolver,
+                                     MOI.ResultCount,
+                                     MOI.TerminationStatus,
+                                     MOI.PrimalStatus,
+                                     MOI.DualStatus,
+                                     MOI.VariablePrimal,
+                                     MOI.ConstraintBasisStatus,
+                                     MOI.ConstraintPrimal,
+                                     MOI.ConstraintDual,
+                                     MOI.ConstraintBasisStatus})
+    return true
+end
+function MOI.get(model::CachingOptimizer, attr::MOI.AbstractModelAttribute)
+    if is_result_attribute(attr)
```

Comment:
How would solver-specific result attributes work? Like `GurobiIIS`?

Reply:
They should implement is_result_attribute.

Reply:
What about using `supports` to decide whether to ask the optimizer or the model cache (with priority to the optimizer)? I can't easily think of a case where this would give the wrong answer, and is much simpler.

Reply:
`supports` is not currently defined for attributes that never appear in `ListOf...AttributesSet` so it does not make sense for result attibutes which are not settable

Reply:
This points to a larger design issue that needs to be resolved before tagging 0.6, but maybe this PR isn't the best place to discuss. It's definitely not ok for solvers to be required to overload functions in MOIU to get correct behavior.

Reply:
Discussion moved to https://github.com/JuliaOpt/MathOptInterface.jl/issues/480 for more visibility